scala/src/com/google/idea/blaze/scala/run/producers/DeployableJarRunConfigurationProducer.java

```diff
@@ -0,0 +1,140 @@
+      return false;
+    }
+
+    File mainObjectFile = RunUtil.getFileForClass(mainObject);
+    if (mainObjectFile == null) {
+      return false;
+    }
+
+    TargetIdeInfo target = findTarget(context.getProject(), mainObjectFile);
+    if (target == null) {
+      return false;
+    }
+
+    Label label = target.key.label;
+    File jarFile = getDeployJarFile(label, context.getProject());
+
+    configuration.setVMParameters("-cp " + jarFile.getPath());
+    configuration.setMainClassName(mainObject.getTruncedQualifiedName());
+    configuration.setModule(context.getModule());
+
+    configuration.putUserData(TARGET_LABEL, label);
+    configuration.setName(mainObject.name());
+    configuration.setNameChangedByUser(true); // don't revert to generated name
+
+    setDeployableJarGeneratorTask(configuration);
+
+    return true;
+  }
+
+  @Override
+  protected boolean doIsConfigFromContext(
+    ApplicationConfiguration configuration,
+    ConfigurationContext context) {
+    ScObject mainObject = getMainObject(context);
+    if (mainObject == null) {
+      return false;
+    }
+
+    File mainObjectFile = RunUtil.getFileForClass(mainObject);
+    if (mainObjectFile == null) {
+      return false;
+    }
+
+    TargetIdeInfo target = findTarget(context.getProject(), mainObjectFile);
+    if (target == null) {
+      return false;
+    }
+
+    if (configuration.getMainClass() == null) {
+      return false;
+    }
+    if (configuration.getVMParameters() == null) {
+      return false;
+    }
+
+    Label label = target.key.label;
+    File jarFile = getDeployJarFile(label, context.getProject());
+
+    return configuration.getMainClass().getQualifiedName().equals(mainObject.getTruncedQualifiedName()) &&
+           configuration.getVMParameters().contains("-cp " + jarFile.getPath());
+  }
+
+  private TargetIdeInfo findTarget(Project project, File sourceFile) {
+    BlazeProjectData projectData =
+      BlazeProjectDataManager.getInstance(project).getBlazeProjectData();
+    if (projectData == null) {
+      return null;
+    }
+
+    Collection<TargetIdeInfo> targets =
+      new FilteredTargetMap(
+        project,
+        projectData.artifactLocationDecoder,
+        projectData.targetMap,
+        t -> t.kind == Kind.SCALA_LIBRARY && t.isPlainTarget())
+        .targetsForSourceFile(sourceFile);
+
+    return Iterables.getFirst(targets, null);
+  }
+
+  private void setDeployableJarGeneratorTask(ApplicationConfiguration config) {
+    Project project = config.getProject();
+    RunManagerEx runManager = RunManagerEx.getInstanceEx(project);
+    runManager.setBeforeRunTasks(config,
+      ImmutableList.of(new GenerateExecutableDeployableJarProviderTaskProvider.Task()));
+  }
+
+  public File getDeployJarFile(Label target, Project project) {
+    WorkspaceRoot root = WorkspaceRoot.fromProject(project);
+    return root.fileForPath(WorkspacePath.createIfValid(
+      String.format("bazel-bin/%s_deploy.jar", target.targetName())));
```

Comment:
This relies on some bazel implementation details (and won't work internally, for example). Can you get this from the BuildResultHelper in the before-run task instead, and fill in the additional run configuration details then?

Reply:
Hmm, I think I have a bit of a chicken-and-egg problem here, because I need to provide the path to .jar in the configuration details prior to my before-task is allowed to run, I'm not quite sure if/how to do this after it had finished (can I even update a configuration object outside of its producer?)

Alternatively, is there a better way to "guestimate" the output jar location? Any helpers to give me the projected output directory?

Reply:
@brendandouglas Additional issue I thought of is the `doIsConfigFromContext` - I would need the jar path to compare with the existing value. So any way of retrieving the output path for the deploy jar before actual execution would be very helpful.
Do you have any suggestions?

Reply:
If you replace 'bazel-bin' by BlazeInfo#blazeBinKey, that will be an improvement.

Otherwise you could also replace the use of ApplicationConfiguration with a custom run configuration type which doesn't require the deploy jar location to be known up front.

This is fine though -- if you like I'll take a look at these things when upstreaming the PR.

Reply:
@brendandouglas if you could look at that upstream, it would be great. I'd really love to get this merged so we can at least start using it.
I'll send a change for the `bazel-bin` string.